runtime/compiler/runtime/JITServerAOTCache.cpp

```diff
@@ -895,6 +1211,220 @@ JITServerAOTCache::writeCache(FILE *f) const
+          (version._jitserverVersion == currentVersion._jitserverVersion);
+   }
+
+// Read an AOT cache snapshot, returning NULL if the cache is ill-formed or
+// incompatible with the running server.
+JITServerAOTCache *
+JITServerAOTCache::readCache(FILE *f, const std::string &name, TR_Memory &trMemory)
+   {
+   if (!JITServerAOTCacheMap::cacheHasSpace())
+      return NULL;
+
+   JITServerAOTCacheHeader header = {0};
+   if (1 != fread(&header, sizeof(JITServerAOTCacheHeader), 1, f))
+      return NULL;
+
+   if (!isCompatibleSnapshotVersion(header._version))
+      return NULL;
+
+   JITServerAOTCache *cache = new (TR::Compiler->persistentGlobalMemory()) JITServerAOTCache(name);
+   if (!cache)
+      return NULL;
+
+   if (!cache->readCache(f, header, trMemory))
+      {
+      cache->~JITServerAOTCache();
+      TR::Compiler->persistentGlobalMemory()->freePersistentMemory(cache);
+      return NULL;
+      }
+
+   return cache;
+   }
+
+bool
+JITServerAOTCache::readCache(FILE *f, const JITServerAOTCacheHeader &header, TR_Memory &trMemory)
+   {
+   _classLoaderMap.reserve(header._numClassLoaderRecords);
+   _classMap.reserve(header._numClassRecords);
+   _methodMap.reserve(header._numMethodRecords);
+   _classChainMap.reserve(header._numClassChainRecords);
+   _wellKnownClassesMap.reserve(header._numWellKnownClassesRecords);
+   _aotHeaderMap.reserve(header._numAOTHeaderRecords);
+   _cachedMethodMap.reserve(header._numCachedAOTMethods);
+
+   _nextClassLoaderId = header._nextClassLoaderId;
+   _nextClassId = header._nextClassId;
+   _nextMethodId = header._nextMethodId;
+   _nextClassChainId = header._nextClassChainId;
+   _nextWellKnownClassesId = header._nextWellKnownClassesId;
+   _nextAOTHeaderId = header._nextAOTHeaderId;
+
+   TR::StackMemoryRegion stackMemoryRegion(trMemory);
+   Vector<AOTCacheClassLoaderRecord *> classLoaderRecords(header._nextClassLoaderId, NULL, stackMemoryRegion);
+   Vector<AOTCacheClassRecord *> classRecords(header._nextClassId, NULL, stackMemoryRegion);
+   Vector<AOTCacheMethodRecord *> methodRecords(header._nextMethodId, NULL, stackMemoryRegion);
+   Vector<AOTCacheClassChainRecord *> classChainRecords(header._nextClassChainId, NULL, stackMemoryRegion);
+   Vector<AOTCacheWellKnownClassesRecord *> wellKnownClassesRecords(header._nextWellKnownClassesId, NULL, stackMemoryRegion);
+   Vector<AOTCacheAOTHeaderRecord *> aotHeaderRecords(header._nextAOTHeaderId, NULL, stackMemoryRegion);
+
+   for (size_t i = 0; i < header._numClassLoaderRecords; ++i)
+      {
+      if (!JITServerAOTCacheMap::cacheHasSpace())
+         return false;
+
+      auto record = AOTCacheClassLoaderRecord::read(f, classLoaderRecords, classRecords, methodRecords,
+                                                    classChainRecords, wellKnownClassesRecords, aotHeaderRecords);
+      if (!record)
+         return false;
+
+      if ((record->data().id() >= classLoaderRecords.size() ||
```

Comment:
IDs should always increase by 1.
Maybe it's better to check that by starting the `for` loop from 1:
`for (size_t i = 1; i <= header._numClassLoaderRecords; ++i)`
and check for inequality inside:
```
if (record->data().id() != i) 
   {
   AOTCacheRecord::free(record);
   return false;
   }
```

Reply:
I would prefer not to bake this assumption into the code. E.g. if we ever implement purging stale methods from the cache, there will be gaps in the record ID sequences. The snapshot header stores both the numbers of records and maximum IDs of each type.

Reply:
ok. Then should we add another test that the position we are trying to write to is not already occupied by another ID?

Reply:
The test is already here: `|| classLoaderRecords[record->data().id()]`. `classLoaderRecords` is initialized to all nulls.

Reply:
Isn't that what the `classLoaderRecords[record->data().id()]` part of test does? Or are you referring to something else?

Reply:
Agreed, I was reading that statement differently.